### ISSUE TITLE
Improve portability by removing hard coded paths

### ISSUE THREAD

**Issue by johnbradley** [hdr-bgnn/minnowTraits#4]
Some of the scripts have hard coded paths. For example:

https://github.com/hdr-bgnn/minnowTraits/blob/686113ec48bff0a990314b23720e299e237871f9/Scripts/selectionCriteriaSegmentedImages.R#L13-L14
Update the scripts removing the hard coded paths.

**Comment by meghalithic** [comment]
@johnbradley what is a way to do that?

**Comment by johnbradley** [comment]
> what is a way to do that?

There are several options. You can use relative paths so that everything is based on the user's current directory. You could also require users to setup a config file with paths to where they want to save/read data. I was going to create a PR with some changes to remove the hard coded paths.

**Comment by johnbradley** [comment]
The **selectionCriteriaSegmentedImages.R** script uses two separate directories for data.

1. a shared project directory
https://github.com/hdr-bgnn/minnowTraits/blob/38038b9644d253a239b3b7839c5954533b31843d/Scripts/selectionCriteriaSegmentedImages.R#L14-L15

2. a user's home directory:
https://github.com/hdr-bgnn/minnowTraits/blob/38038b9644d253a239b3b7839c5954533b31843d/Scripts/selectionCriteriaSegmentedImages.R#L58

Do we want to continue to store the data in two locations like this? Or would it be better to store all the data in one location?

**Comment by meghalithic** [comment]
probably in one location, but the issue is how Snakemake works (@thibaulttabarin). Perhaps a meeting to discuss would be best.

**Comment by johnbradley** [comment]
@megbalk I'm updating a branch with changes to fix this issue and #3. Do you know roughly how long minnowSelectionImageQualityMetadata.R and selectionCriteriaSegmentedImages.R take to run?

**Comment by meghalithic** [comment]
@johnbradley minnowSelectionImageQualityMetadata.R at most takes 15 min., and selectionCriteriaSegmentedImages.R is less.

**Comment by meghalithic** [comment]
@johnbradley how does a config file work? I am worried that this would be difficult for biologists like me to use.

**Comment by johnbradley** [comment]
> how does a config file work? I am worried that this would be difficult for biologists like me to use.

For snakemake it is a yaml file. Here is one from a branch I am working on:

https://github.com/hdr-bgnn/minnowTraits/blob/c2d3a195d624b22fef638713214a5f284f9a8870/config/config.yaml#L1-L8

Typically users just specify the input files in the config file.
Since the output files have embedded timestamps I decided to put them here as well.
Users can also pass the each option from the command line if that is preferable.

**Comment by meghalithic** [comment]
I will make a file for relative paths to call. I will also will add file names to the scripts.

**Comment by meghalithic** [comment]
@johnbradley how do I make a relative path of the outputs of the workflow?

**Comment by johnbradley** [comment]
> @johnbradley how do I make a relative path of the outputs of the workflow?

@megbalk I think you already have relative paths for the output files. For example:

https://github.com/hdr-bgnn/Minnow_Traits/blob/1019ec9fc279288a0587c706fc8f208a71964f8e/Scripts/minnowSelectionImageQualityMetadata.R#L177-L182

You will probably want to add a directory prefix for example something like:
```
write.csv(images.minnows.10, "results/minnow.filtered.from.imagequalitymetadata_25Jul2022.csv", row.names = FALSE) 
```

**Comment by meghalithic** [comment]
@johnbradley check out the branch mab-issue-4, then Scripts/Morphology_Analyses.R (lines 32-50) or Scripts/Selection_Criteria_Segmented_Images.R (lines 35-50)

I need to be in the OSC folder to turn all the OSC json files into a dataframe, then return to my repository to write the file. I am unsure how to not hard code that.

**Comment by johnbradley** [comment]
> I need to be in the OSC folder to turn all the OSC json files into a dataframe, then return to my repository to write the file. I am unsure how to not hard code that.

I think there are two issues here:
1) hard coding the directory
2) setwd logic

##  hard coding the directory
Use a relative path instead. I am assuming these are the files generated by the BGNN_Snakemake repo.
We can setup the pipeline to create these json files wherever you would like.
So for example they could be written to a relative path like so: `results/Burress/Morphology/Presence`.

##  setwd logic
There is no need to keep changing your working directory. I have often encountered problems when writing logic that changes the working directory. It is really easy to forget to change back and end up breaking things. The R functions for working with files will work fine without being within the directory containing the files. There is a bug in `json_df()` though.

#  Changes
Here is one of the examples above:
https://github.com/hdr-bgnn/Minnow_Traits/blob/f7221b0dba0b3e279505e83cdbafc906c1fb431a/Scripts/Morphology_Analyses.R#L32-L51

Removing setwd and using a relative path would look like so:
```
m.files <- list.files(path = file.path("results/Burress/Morphology/Presence", measure), pattern = '*.json') 
  
# turn into csv 
measure.df <- lapply(m.files, json_df) %>% #list of dataframes 
   dplyr::bind_rows() #rbind to turn into single dataframes 
...
# write data frame to Results directory 
 write.csv(measure.df,  
           file = file.path(results, paste0("measure.df_", Sys.Date(), ".csv")), 

```

You may need this change to `json_df` to remove the directory after removing the `setwd()` logic.
https://github.com/hdr-bgnn/Minnow_Traits/blob/8bdee8fa0a111259259a36790d6de4e7840f9b1d/Scripts/selectionCriteriaSegmentedImages.R#L33-L40

You can see similar fixes I created for earlier versions of the scripts in this PR: https://github.com/hdr-bgnn/Minnow_Traits/pull/16

**Comment by meghalithic** [comment]
let's discuss at our next meeting. I think there is something being missed.

**Comment by meghalithic** [comment]
* we need everything in one directory, so maybe need to run Snakemake and have results from BGNN_Snakemake go into this repo 
* create a config and Snakemake file in minnow_traits and then have it call BGNN_Snakemake (so don't need to check out other repo) that has tag with version number [see branch 8bdee8fa0a]
* put all outputs of BGNN_Snakemake to .gitignore (so can call locally, but won't be committed to GitHub); will upload to TDI data commons [@johnbradley help]
* on BGNN_Snakemake, add section for "To use on your project / repo": create config file, here is how to create it and what to include, create a snakefile, here is how to create it and what to include, here's are recommendation for how to structure your project's repo
* @johnbradley and @megbalk will work to resolve the redundancies between mab-isssue-4 and 8bdee8fa0a branches, specifically paths.R, dependencies.R, and config.yml; next they will look at the PR #25 
* take code for massaging data into own script and source it
* need archival of everything on dryad (inputs and outputs from automated process)

**Comment by meghalithic** [comment]
@johnbradley I'm not sure how to tell the R Script to use the file name from the Snakefile

**Comment by meghalithic** [comment]
@johnbradley also, how does snakemake know to run R and do I have to source dependencies and paths in every script?

**Comment by johnbradley** [comment]
> @johnbradley I'm not sure how to tell the R Script to use the file name from the Snakefile

@megbalk I added a commit with the **config/config.yaml** file to the mab-issue-4 branch.
I think we will need to add a library to read the yaml file.
I tried out the `yaml` package and was able to read it like so:
```
> install.packages("yaml")
...
> config = yaml::read_yaml("config/config.yaml")
> config$burress
[1] "Files/Previous Fish Measurements - Burress et al. 2016.csv"
> config$image_metadata
[1] "Files/Image_Metadata_v1_20211206_151152.csv"
```

**Comment by johnbradley** [comment]
> @johnbradley also, how does snakemake know to run R

This is a great question. 
In Snakefile we specify the `Rscript` command to run under the `shell` section of a rule.
Here is a simplified version of a rule:
```
rule select_minnow_images:
    shell:
        "Rscript Scripts/minnowSelectionImageQualityMetadata.R"
    ...
```
By default snakemake assumes all scripts required (R and dependencies) are installed and snakemake can just run the commands in **shell**. To increase portability of this workflow I will create a Docker container with R and all the dependencies installed. Then we update the Snakefile telling it the container to use for each command we run.

---

> do I have to source dependencies and paths in every script?

I would not include dependencies.R within every script. Having the script automatically install dependencies would be really surprising. It would also limit users from trying newer versions of the packages when experimenting. Instead we should document in the README or somewhere similar that users who want to run the R scripts need to run dependencies.R first. We can use dependencies.R to build our docker container.

For paths.R you only need to include it in your R scripts that need the variables it sets up. So anywhere that needs to know the input and output filenames.

**Comment by meghalithic** [comment]
Also, some of the dependencies require to click a pop up and to type in "yes". I'm not sure how to do that automated.

**Comment by meghalithic** [comment]
> > @johnbradley I'm not sure how to tell the R Script to use the file name from the Snakefile
> 
> @megbalk I added a commit with the **config/config.yaml** file to the mab-issue-4 branch. I think we will need to add a library to read the yaml file. I tried out the `yaml` package and was able to read it like so:
> 
> ```
> > install.packages("yaml")
> ...
> > config = yaml::read_yaml("config/config.yaml")
> > config$burress
> [1] "Files/Previous Fish Measurements - Burress et al. 2016.csv"
> > config$image_metadata
> [1] "Files/Image_Metadata_v1_20211206_151152.csv"
> ```

By library do you mean an R package to read yaml files?

**Comment by johnbradley** [comment]
> By library do you mean an R package to read yaml files?

Yes.

**Comment by johnbradley** [comment]
> Also, some of the dependencies require to click a pop up and to type in "yes". I'm not sure how to do that automated.

What prompt would that be? Perhaps the `Do you want to restart R prior to install?` in RStudio. If so that should not be an issue since we will not be using RStudio just R. I hope to just run `Rscript dependencies.R` to install the requirements into the docker container.

**Comment by johnbradley** [comment]
@megbalk I pushed a few commits to the `mab-issue-4` branch.
- Removed Library directory 95fc4cb1cd22c93b3128e2217b7ed419ddc231a0
- Removed sampling.df.IQM.csv output file c8f01993103e0dc00655e919bf54bfdc45f3c40e
- Fixed some pathing problems 72e40a5bed464a88fa1c30b267f445f3aaec9cca
- Fixed dependencies.R to be more portable and include required paths.R 63d5ee866d4cb6bbe28c4d703356e6eb264071d1
-  Adds minimal snakemake workflow 33cc290684942642da5dc0846ef91c324fc2b9be

I've tested these changes locally from the command line using `Rscript`.
I still need to test them on OSC.

---

I was wondering if I should add the following to `init.R`:
```
.libPaths(c(.libPaths(), file.path(getwd(), library)))
```
I don't see how "Library/" gets set up for use with Minnow_Selection_Image_Quality_Metadata.R otherwise.

**Comment by meghalithic** [comment]
That is for all the dependencies, so it shouldn't be needed for Minnow_Selection_Image_Quality_Metadata.R

**Comment by johnbradley** [comment]
On OSC I was able to install the dependencies into `Library` like so:
```
module load cmake/3.20.5 
module load R/4.2.1-gnu11.2
Rscript dependencies.R 
```
I included the cmake module after `nloptr` (one of the nested depenencies)  failed with 
```
CMake Error at CMakeLists.txt:15 (cmake_minimum_required):
   CMake 3.2 or higher is required.  You are running version 2.8.12.2
```

**Comment by johnbradley** [comment]
When first trying to run on OSC after installing the dependencies I had this problem:
```
$ Rscript Scripts/Minnow_Selection_Image_Quality_Metadata.R 
Error in library(dplyr) : there is no package called ‘dplyr’
Execution halted
```

I was able to get the script running like so:
```
R_LIBS_USER=Library Rscript Scripts/Minnow_Selection_Image_Quality_Metadata.R
```
This sets the [R_LIBS_USER environment variable](https://stat.ethz.ch/R-manual/R-devel/library/base/html/libPaths.html) to the `Library` - the directory created by `dependencies.R`.

**Comment by johnbradley** [comment]
@megbalk I added some instructions to the README in the branch for [how to run the snakemake workflow](https://github.com/hdr-bgnn/Minnow_Traits/tree/mab-issue-4#running-the-workflow). When you get a chance could you try following them to run the workflow on OSC? You may want to delete Results/burress.minnow.sp.filtered.from.iqm.csv, Results/minnow.filtered.from.iqm.csv, and Results/sampling.df.IQM.csv before running the workflow if these files still exist in your repo.

Thanks.

**Comment by meghalithic** [comment]
@johnbradley I get the following error for module load cmake/3.20.5: 
`Lmod has detected the following error:  The following module(s) are unknown: "cmake/3.20.5"
`

**Comment by johnbradley** [comment]
> I get the following error for module load cmake/3.20.5:
Lmod has detected the following error: The following module(s) are unknown: "cmake/3.20.5"

Are you on Pitzer? The following works fine for me there:
```
[jbradley@pitzer-login02 ~]$ module load cmake/3.20.5
[jbradley@pitzer-login02 ~]$ cmake --version
cmake version 3.20.5
```

**Comment by meghalithic** [comment]
I'm on owens, because the other one is busy.
does it matter which one?
Also, I am in the minnow traits cloned repository in the shared team folder, do I need to be on the login node?

**Comment by johnbradley** [comment]
It looks like owens and pitzer do not share the same modules. This was surprising to me.
```
[jbradley@owens-login01 ~]$ module avail cmake

------------------------------------------- /usr/local/share/lmodfiles/Core -------------------------------------------
   cmake/3.6.1    cmake/3.7.2    cmake/3.11.4    cmake/3.16.5    cmake/3.17.2 (D)
```

I don't think the version of cmake really matters as long as you have one.
You can use this command instead: 
```
module load cmake/3.17.2
```
or
```
module load cmake
```
If you leave off the version part of the module name the module command will use the default one.

**Comment by johnbradley** [comment]
> does it matter which one?

The code should work on both clusters. Really it should work on any Slurm cluster that has singularity installed.

> Also, I am in the minnow traits cloned repository in the shared team folder, do I need to be on the login node?

As long as you are in the root directory of the cloned repo you should be good. With OSC using the login node to install software is their recommendation. I would expect everything to still work fine even if you are in an interactive session(and thereby no longer on the login node).

**Comment by megbalk** [comment]
interesting, i wonder how we'll make it generalizable for everyone then....?

**Comment by meghalithic** [comment]
maybe the readme should have:

`module load cmake
`

**Comment by johnbradley** [comment]
Once we switch to using a container these steps can be removed. I held off on creating the container to simplify things.

**Comment by meghalithic** [comment]
@johnbradley things are running!

I added more documentation (mostly for me) to the readme

**Comment by meghalithic** [comment]
@johnbradley I don't think it worked...

**Comment by johnbradley** [comment]
> @johnbradley I don't think it worked...

Can you elaborate more? Under `.../BGNN/Minnow_Traits/Results` I see the three output files created yesterday:
```
-rw-r--r--  1 balkm PAS2136 2887990 Sep  1 17:18 minnow.filtered.from.iqm.csv
-rw-r--r--  1 balkm PAS2136  204526 Sep  1 17:18 burress.minnow.sp.filtered.from.iqm.csv
drwxr-xr-x  3 balkm PAS2136    4096 Sep  1 17:18 .
-rw-r--r--  1 balkm PAS2136     536 Sep  1 17:18 sampling.df.IQM.csv
```
It looks like snakemake ran to completion:
```
$ tail -n 4 .snakemake/log/2022-09-01T165214.578988.snakemake.log 
[Thu Sep  1 17:18:13 2022]
Finished job 0.
2 of 2 steps (100%) done
Complete log: .snakemake/log/2022-09-01T165214.578988.snakemake.log
```

**Comment by meghalithic** [comment]
@johnbradley datasets have been added to the Data Commons!

**Comment by johnbradley** [comment]
@megbalk When you get a chance could you try running the pipeline on OSC again after pulling the latests changes? I pushed a commit that will create a subset of the morphology json files(https://github.com/hdr-bgnn/Minnow_Traits/commit/52b7754a097ed8b396cfd3d438c7d587f0aba668). I changed the workflow to process a subset so it will not take a really long time to run. The output files from BGNN_Snakemake will be under a "segmentation" subdirectory.
On the login node just run the workflow as usual:
```
sbatch run-workflow.sh 
```
Since the BGNN_Snakemake workflow uses singularity I enabled that in `run-workflow.sh`. The workflow will take a little while to download those containers, but there will be messages in the slurm logs.

The workflow creates morphology JSON files based on the "Minnow_Filtered" CSV file. Please let me know if we need to create morphology data for the images from the "Burress_Minnow_Filtered" file instead.

**Comment by meghalithic** [comment]
@johnbradley it worked!

**Comment by johnbradley** [comment]
Good to hear @megbalk. I noticed there are quite a few R scripts under `Scripts` now.  Which ones need to be included in the workflow? Thanks.

**Comment by johnbradley** [comment]
I pushed another commit adding `Selection_Criteria_Segmented_Images.R` to `workflow/Snakefile`.
I also finished removing the last hard coded paths in that script.
See changes here: https://github.com/hdr-bgnn/Minnow_Traits/commit/b0e3b46cd767a20f23443b26ba0d8354612f32cd

@megbalk When you get a chance could you try out these changes? Thanks.

**Comment by meghalithic** [comment]
@johnbradley do you mean re-run it? 

Which part of the workflow are you referring to for the scripts?

**Comment by johnbradley** [comment]
> @johnbradley do you mean re-run it?
Which part of the workflow are you referring to for the scripts?

Run the whole workflow on OSC. Pull the latests changes then run 
```
sbatch run-workflow.sh 
```

**Comment by meghalithic** [comment]
@johnbradley done...was it supposed to be that quick?

**Comment by johnbradley** [comment]
> @johnbradley done...was it supposed to be that quick?

It looks like Snakemake created the appropriate png and csv files for Selection_Criteria_Segmented_Images.R. 
Snakemake skipped the early steps since the input files didn't change. Snakemake just just ran `Selection_Criteria_Segmented_Images.R`. Another reason for the speed is that we are only processing 10 morphology JSON files right now. 

If you are comfortable with the changes I made here: https://github.com/hdr-bgnn/Minnow_Traits/commit/b0e3b46cd767a20f23443b26ba0d8354612f32cd I can push a commit to remove the limit and setup Snakemake to use Slurm to run the jobs in parallel.

**Comment by meghalithic** [comment]
looks good! what why do you select [1] for "dfs$Presence_Absence_Matrix[1]"?

**Comment by johnbradley** [comment]
> why do you select [1] for "dfs$Presence_Absence_Matrix[1]"?

I was just following the preceding logic for other filenames:
https://github.com/hdr-bgnn/Minnow_Traits/blob/b0e3b46cd767a20f23443b26ba0d8354612f32cd/Scripts/Data_Manipulation.R#L14-L18

I figured this was how one used the output of `yaml::read_yaml()`:
https://github.com/hdr-bgnn/Minnow_Traits/blob/b0e3b46cd767a20f23443b26ba0d8354612f32cd/Scripts/Data_Manipulation.R#L6

**Comment by johnbradley** [comment]
After playing with `yaml::read_yaml` for a bit I think all of the `[1]` parts are unnecessary:
```
> dfs <- yaml::read_yaml(file = "config/config.yaml") 
> dfs$Image_Metadata
[1] "Files/Image_Metadata_v1_20211206_151152.csv"
> dfs$Image_Metadata[1]
[1] "Files/Image_Metadata_v1_20211206_151152.csv"
> class(dfs$Image_Metadata)
[1] "character"
> class(dfs$Image_Metadata[1])
[1] "character"
> typeof(dfs$Image_Metadata)
[1] "character"
> typeof(dfs$Image_Metadata[1])
[1] "character"
```
@megbalk Any objection to me removing the `[1]` bits?

**Comment by meghalithic** [comment]
go for it!

**Comment by johnbradley** [comment]
@megbalk I added changes to use Slurm and removed the limit and setup Snakemake to use Slurm to run the jobs in parallel. I added a commit that removes the extra `[1]` logic.

Here are the Slurm changes: https://github.com/hdr-bgnn/Minnow_Traits/commit/dc0329b64bb012e5245702ea38b5bea8502b8861

Here are changes that simplify Data_Manipulation.R: https://github.com/hdr-bgnn/Minnow_Traits/commit/efc47b46bb51f8d0b0d1b12bcf369aa86bbda1a9

On OSC Pull the latest changes and run the workflow again. Same as before:
```
sbatch run-workflow.sh 
```
Under the hood Snakmake will create a sbatch job for each step. Currently `run-workflow.sh` configures Snakemake to run up to 20 jobs in parallel. This should take a while even with running code in parallel since I removed the limit for the number of morphology JSON files created.

**Comment by meghalithic** [comment]
hmmm it hit a time limit (see slurm-122882402.out)

**Comment by johnbradley** [comment]
> hmmm it hit a time limit (see slurm-122882402.out)

I tested slurm with the 10 file limit still in place so I didn't see this issue.
We need to set the timeout parameter for the main sbatch job (run-workflow.sh).
The sbatch command has a parameter `--time=<time_limit>` to specify this.
I will add this flag to run-workflow.sh using the `#SBATCH` comment so we don't need to remember it.

**Comment by johnbradley** [comment]
@megbalk I pushed a change to set a 3 hour time limit for run-workflow.sh:
https://github.com/hdr-bgnn/Minnow_Traits/blob/1a1d565546cf7564513d9a1200116f6cae341477/run-workflow.sh#L4

I determined 3 hours by adding 30 minutes for your scripts to the 2hr30min time limit used in BGNN_Snakemake.

Could you pull the changes and run `sbatch run-workflow.sh` again?

---

There is a possibility that you will get an error about snakemake saying the [directory is locked](https://snakemake.readthedocs.io/en/stable/project_info/faq.html#how-does-snakemake-lock-the-working-directory) if snakemake didn't get enough time to cleanup. If you do have this error it will say to run snakemake with `--unlock`. Since the command line arguments passed to `run-workflow.sh` are passed to snakemake you can just do the following:
```
sbatch run-workflow.sh --unlock
```

**Comment by meghalithic** [comment]
ok, i think it worked (slurm-12897163.out)

**Comment by johnbradley** [comment]
> ok, i think it worked (slurm-12897163.out)

When run with `--unlock` snakemake doesn't run the pipeline. It just unlocks the working directory. Based on the slurm log it unlocked the working directory. So now you should be good to run the pipeline:
```
sbatch run-workflow.sh
```

**Comment by meghalithic** [comment]
oh! ok rerunning it, it is job 12898119

**Comment by meghalithic** [comment]
ok now done!

**Comment by johnbradley** [comment]
@megbalk From what I can tell it looks like it worked correctly. Could you review the output files in Results to see if the file contents look good. There are still some improvements we could make to the Snakemake workflow, but as far as hard coded paths I think we are good. If everything looks good in Results how do you feel about creating a PR for this branch and merging it into main?

**Comment by johnbradley** [comment]
Fixed by #28
